Kernel/Devices/Audio/IntelHDA/Controller.cpp

```diff
@@ -0,0 +1,304 @@
+
+static constexpr u32 frame_delay_in_microseconds(u32 frames)
+{
+    // NOTE: the link operates at this _fixed_ frequency and is independent of the streams' rates.
+    constexpr u32 link_frame_frequency_hz = 48'000;
+
+    // 2.2: Streams and Channels
+    // A new frame starts exactly every 20.83 μs, corresponding to the common 48-kHz sample rate.
+    VERIFY(frames <= 4294);
+    return frames * 1'000'000u / link_frame_frequency_hz + 1u;
+}
+
+UNMAP_AFTER_INIT ErrorOr<NonnullLockRefPtr<Controller>> Controller::create(PCI::DeviceIdentifier const& pci_device_identifier)
+{
+    auto controller_io_window = TRY(IOWindow::create_for_pci_device_bar(pci_device_identifier, PCI::HeaderType0BaseRegister::BAR0));
+
+    auto intel_hda = TRY(adopt_nonnull_lock_ref_or_enomem(new (nothrow) Controller(pci_device_identifier, move(controller_io_window))));
+    TRY(intel_hda->initialize());
+    return intel_hda;
+}
+
+UNMAP_AFTER_INIT Controller::Controller(PCI::DeviceIdentifier const& pci_device_identifier, NonnullOwnPtr<IOWindow> controller_io_window)
+    : PCI::Device(const_cast<PCI::DeviceIdentifier&>(pci_device_identifier))
+    , m_controller_io_window(move(controller_io_window))
+{
+}
+
+UNMAP_AFTER_INIT Controller::~Controller() = default;
+
+UNMAP_AFTER_INIT ErrorOr<void> Controller::initialize()
+{
+    // Enable DMA
+    PCI::enable_bus_mastering(device_identifier());
+
+    // 3.3.3, 3.3.4: Controller version
+    auto vmin = m_controller_io_window->read8(ControllerRegister::VersionMinor);
+    auto vmaj = m_controller_io_window->read8(ControllerRegister::VersionMajor);
+    dmesgln_pci(*this, "Intel High Definition Audio specification v{}.{}", vmaj, vmin);
+    if (vmaj != 1 || vmin != 0)
+        return ENOTSUP;
+
+    // 3.3.2: Read capabilities
+    u16 capabilities = m_controller_io_window->read16(ControllerRegister::GlobalCapabilities);
+    dbgln_if(INTEL_HDA_DEBUG, "Controller capabilities:");
+    m_number_of_output_streams = capabilities >> 12;
+    m_number_of_input_streams = (capabilities >> 8) & 0xf;
+    bool is_64_bit_addressing_supported = (capabilities & 0x1) > 0;
+    dbgln_if(INTEL_HDA_DEBUG, "├ Number of output streams: {}", m_number_of_output_streams);
+    dbgln_if(INTEL_HDA_DEBUG, "├ Number of input streams: {}", m_number_of_input_streams);
+    dbgln_if(INTEL_HDA_DEBUG, "└ 64-bit addressing supported: {}", is_64_bit_addressing_supported ? "yes" : "no");
+    if (m_number_of_output_streams == 0)
+        return ENOTSUP;
+    if (!is_64_bit_addressing_supported && sizeof(FlatPtr) == 8)
+        return ENOTSUP;
+
+    // Initialize the controller
+    u32 global_control = m_controller_io_window->read32(ControllerRegister::GlobalControl);
+    global_control |= GlobalControlFlag::ControllerReset;
+    global_control &= ~GlobalControlFlag::AcceptUnsolicitedResponseEnable;
+    m_controller_io_window->write32(ControllerRegister::GlobalControl, global_control);
+    do {
+        microseconds_delay(21);
```

Comment:
Why do you delay read by 21 microseconds here? a comment on this would be very nice.

Reply:
The 21 microseconds delay roughly corresponds to the duration of 1 link frame, so I changed that value to `frame_delay_in_microseconds(1)` to make that abundantly clear.

However, the value itself is not specced or relevant: resetting the controller can take an arbitrary time. I've introduced a new helper function `wait_until` that takes a timeout value into account. An arbitrary (and commented as such) controller timeout of 300 milliseconds was selected.